src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,164 @@
+                + "existing or negative index");
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to set at nonexistent or negative index");
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        T oldValue = nodeAtIndex.value;
+        nodeAtIndex.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexIsValid(index, size - 1, "Trying to remove at nonexistent or negative index");
+        return unlink(findNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        while (temp != null) {
+            if (temp.value == object || temp.value != null && temp.value.equals(object)) {
+                unlink(temp);
+                return true;
+            }
+            temp = temp.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> previous;
+        private Node<T> next;
+
+        public Node(Node<T> previous, T value, Node<T> next) {
+            this.previous = previous;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private void checkIndexIsValid(int index, int size, String message) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(message);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        return index < size / 2 ? findFromHead(index) : findFromTail(index);
+    }
+
+    private Node<T> findFromTail(int index) {
+        int count = size - 1;
+        Node<T> searchResult = tail;
+        while (count != index) {
+            searchResult = searchResult.previous;
+            count--;
+        }
+        return searchResult;
+    }
+
+    private Node<T> findFromHead(int index) {
+        int count = 0;
+        Node<T> searchResult = head;
+        while (count < index) {
+            searchResult = searchResult.next;
+            count++;
+        }
+        return searchResult;
+    }
+
+    private T unlink(Node<T> temp) {
+        Node<T> previous = temp.previous;
+        Node<T> next = temp.next;
+        T value = temp.value;
+        if (previous == null && next == null) {
+            head = null;
+            tail = null;
+        } else if (previous == null) {
+            next.previous = null;
+            head = next;
+        } else if (next == null) {
+            previous.next = null;
+            tail = temp.previous;
+        } else {
+            previous.next = next;
+            next.previous = previous;
+        }
```

Comment:
let's try to simplify this a bit

first check if `next` is null - write `if-else` construction

then check if `previous `is null - write another `if-else` construction

let's also set node's value to null in the end